### PR TITLE
Use non APIPA when assigning IP Address

### PR DESCRIPTION
**Pull request by jbvmio**
**What this PR does**:
Returns the first non Automatic Private IP if possible. Will return APIP as last resort.

**Which issue(s) this PR fixes**:
Fixes or addresses #4014

### CODE REVIEW

**Review by jtlisi — COMMENTED** [review]
Thank you for the contribution, if you get a chance can you look into adding a DCO signature to your commits. Also I'd like to make a suggestion.

**suggestion:** Instead of using two label breaking loops, consider pushing logic in the `filterIPs` for loop into a separate function. This will allow for two things:

1. The code can be rewritten to not use label breaks
2. It will be easy to add unit tests to the new function that replaces the `filterIPs` loop.

@jbvmio WDYT?

**Comment by jbvmio** [comment]
Sounds good to me @jtlisi, appreciate the feedback.
The requested changes have been made, let me know how it looks when you get a chance.

**Review by jtlisi — COMMENTED** [review]
@jbvmio This looks good to merge! Can you add a change log entry?

**Comment by jbvmio** [comment]
@jtlisi Change log entry added, thanks!

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 ## master / unreleased
 
+* [ENHANCEMENT] Ingester/Query-Frontend: Avoid using automatic private IPs (APIPA) upon assignment. APIPA still used as last resort with logging indicating usage. #4032
```
Comment (jtlisi) [inline CHANGELOG.md:5]: @jbvmio Would you mind moving this entry down so it is above the other `ENHANCEMENT` entries?

```suggestion
* [ENHANCEMENT] Ingester/Query-Frontend: Avoid using automatic private IPs (APIPA) upon assignment. APIPA still used as last resort with logging indicating usage. #4032
```

**Review by jtlisi — APPROVED** [review]
LGTM

File: CHANGELOG.md (hunk 1 above)
Comment (jbvmio) [reply CHANGELOG.md]: No problem, done.

File: pkg/util/net.go [hunk 2]
```diff
@@ -27,16 +28,33 @@ func GetFirstAddressOf(names []string) (string, error) {
 			level.Warn(util_log.Logger).Log("msg", "no addresses found for interface", "inf", name, "err", err)
 			continue
 		}
+		ipAddr = filterIPs(addrs)
```
[the remark this answers is not in the document]
Comment (jbvmio) [reply pkg/util/net.go]: Nice catch @pstibrany with an easy enough fix for prevention. Updated accordingly.

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,7 +6,9 @@
 * [CHANGE] Alertmanager: allowed to configure the experimental receivers firewall on a per-tenant basis. The following CLI flags (and their respective YAML config options) have been changed and moved to the limits config section: #4143
   - `-alertmanager.receivers-firewall.block.cidr-networks` renamed to `-alertmanager.receivers-firewall-block-cidr-networks`
   - `-alertmanager.receivers-firewall.block.private-addresses` renamed to `-alertmanager.receivers-firewall-block-private-addresses`
+* [CHANGE] Distributor: Added ring status section in the admin page #4151
 * [CHANGE] Change default value of `-server.grpc.keepalive.min-time-between-pings` to `10s` and `-server.grpc.keepalive.ping-without-stream-allowed` to `true`. #4168
+* [FEATURE] Alertmanager: Added rate-limits to email notifier. Rate limits can be configured using `-alertmanager.email-notification-rate-limit` and `-alertmanager.email-notification-burst-size`. These limits are applied on individual alertmanagers. Rate-limited email notifications are failed notifications. It is possible to monitor rate-limited notifications via new `cortex_alertmanager_notification_rate_limited_total` metric. #4135
```
Comment (pstibrany) [inline CHANGELOG.md:11]: This should not be here. Master already has updated version of this changelog entry.

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,7 +6,9 @@
 * [CHANGE] Alertmanager: allowed to configure the experimental receivers firewall on a per-tenant basis. The following CLI flags (and their respective YAML config options) have been changed and moved to the limits config section: #4143
   - `-alertmanager.receivers-firewall.block.cidr-networks` renamed to `-alertmanager.receivers-firewall-block-cidr-networks`
   - `-alertmanager.receivers-firewall.block.private-addresses` renamed to `-alertmanager.receivers-firewall-block-private-addresses`
+* [CHANGE] Distributor: Added ring status section in the admin page #4151
```
Comment (pstibrany) [inline CHANGELOG.md:9]: This should not be here. Master already has updated version of this changelog entry.

File: pkg/util/net.go [hunk 5]
```diff
@@ -11,13 +12,13 @@ import (
 
 // GetFirstAddressOf returns the first IPv4 address of the supplied interface names.
 func GetFirstAddressOf(names []string) (string, error) {
```
Comment (pstibrany) [inline pkg/util/net.go:14]: Let's also update the comment for this function and mention handling IP Addresses in APIPA range.

File: CHANGELOG.md [hunk 6]
```diff
@@ -19,6 +21,7 @@
   * `cortex_alertmanager_state_persist_failed_total`
 * [ENHANCEMENT] Blocks storage: support ingesting exemplars.  Enabled by setting new CLI flag `-blocks-storage.tsdb.max-exemplars=<n>` or config option `blocks_storage.tsdb.max_exemplars` to positive value. #4124
 * [ENHANCEMENT] Distributor: Added distributors ring status section in the admin page. #4151
+* [ENHANCEMENT] Ingester/Query-Frontend: Avoid using automatic private IPs (APIPA) upon assignment. APIPA still used as last resort with logging indicating usage. #4032
```
Comment (pstibrany) [inline CHANGELOG.md:24]:
```suggestion
* [ENHANCEMENT] Ring, query-frontend: Avoid using automatic private IPs (APIPA) when discovering IP address from the interface during the registration of the instance in the ring, or by query-frontend when used with query-scheduler. APIPA still used as last resort with logging indicating usage. #4032
```

**Review by pstibrany — APPROVED** [review]
Thank you. Let's please remove extra changelog entries that are not related to this PR before merging.

File: pkg/util/net.go (hunk 5 above)
Comment (jbvmio) [reply pkg/util/net.go:14]: Sounds good, updated.

File: CHANGELOG.md (hunk 6 above)
Comment (jbvmio) [reply CHANGELOG.md]: Looks good.

**Comment by pstibrany** [comment]
Thank you for addressing my feedback.

**Pull request merged**